Userland/Libraries/LibGfx/TextLayout.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2021, sin-ack <sin-ack@protonmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "TextLayout.h"
+
+namespace Gfx {
+
+// HACK: We need to point to some valid memory with Utf8Views.
+char const s_the_newline[] = "\n";
+
+IntRect TextLayout::bounding_rect(TextWrapping wrapping, int line_spacing) const
+{
+    auto lines = wrap_lines(TextElision::None, wrapping, line_spacing, FitWithinRect::No);
+    if (!lines.size()) {
+        return {};
+    }
+
+    IntRect bounding_rect = {
+        0, 0, 0, static_cast<int>((lines.size() * (m_font->glyph_height() + line_spacing)) - line_spacing)
+    };
+
+    for (auto& line : lines) {
+        auto line_width = m_font->width(line);
+        if (line_width > bounding_rect.width())
+            bounding_rect.set_width(line_width);
+    }
+
+    return bounding_rect;
+}
+
+Vector<String, 32> TextLayout::wrap_lines(TextElision elision, TextWrapping wrapping, int line_spacing, FitWithinRect fit_within_rect) const
+{
+    Vector<Utf8View> words;
+
+    Optional<size_t> start_byte_offset;
+    size_t current_byte_offset = 0;
+    for (auto it = m_text.begin(); !it.done(); ++it) {
+        current_byte_offset = m_text.iterator_offset(it);
+
+        switch (*it) {
+        case '\n':
+        case '\r':
+        case '\t':
+        case ' ': {
+            if (start_byte_offset.has_value())
+                words.append(m_text.substring_view(start_byte_offset.value(), current_byte_offset - start_byte_offset.value()));
+            start_byte_offset.clear();
+
+            // FIXME: is there anyway we can get Mac newlines (\r) in text?
```

Comment:
I don't understand what this FIXME defers.
"FIXME: Add support for mac-style newlines (\r)"? not sure why that would be necessary or ever happen.

Reply:
Very good question. I don't remember writing this.